### ISSUE TITLE
Semantic tokens: Allow extensions to provide tokens eagerly

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#86168]
Testing #85976

I mention in https://github.com/microsoft/vscode/issues/86165 how the provide tokens call is only triggered after typing has stopped for a while. If an extension wants to "go all in" and provide edits to that function, I would expect it to be able to react immediately following `onDidChangeTextDocument` which are fired on every change. This will often just result in just increasing the size of a token which is a very minimal change.

For example:

```ts
vscode.workspace.onDidChangeTextDocument(e => {
  // ...
  // The extension evaluated quickly that the symbol's length increased by one, so
  // it can update immediately. If this was costly it could defer to a regular provide call
  const edits = new SemanticTokensEdits([new SemanticTokensEdit(line * 5 + 2, 1, [ oldLength + 1 ])]);
  vscode.languages.updateTokens(edits);
});
```

**Comment by alexandrudima** [comment]
Thanks for your feedback!

The editor is already adjusting tokens in that cheap way on the renderer process. On all text model edits, the semantic tokens get adjusted by the editor. So pressing <kbd>Enter</kbd> will move all of the ones below down, typing will move all of them on the same line to the left... It is possible that a language specific implementation would do better, but then perhaps we can look into configuring the implementation of the editor. For example, right now, there is some smartness where semantic tokens will "grow" when typing at their end, but only when typing a word character. I thought this might work well for most programming languages, but might not work well for others. But I would revisit this once we implement semantic tokens providers in 2-3 languages and we notice some kind of pattern where the current generic adjusting logic is not working well.

We have decided to go for a "pull" model with the idea that this call is one which can be costly and it is the editor that can establish when it's a good time to make it. The big picture thinking is that a language server is mostly written to serve one request after another, i.e. IMHO I don't believe there are any language servers that "reorder" requests. For example, by having the editor decide when to make the request, we could give a higher priority to suggestions on typing, since those are far more useful than semantic tokens when typing. Only after we trigger the suggest request, would we then trigger the semantic tokens request... This would hopefully result in the suggest request getting first served by the language server... We could also measure the time it took the semantic tokens provider to reply and adjust the debounce time dynamically. Right now, the debounce time is hard-coded at 500ms after the last text edit and that might be a bit high.

So, TL;DR I think the pull model is a good model for semantic tokens, and we could improve the synchronous on-type token adjustment in the editor if we find it is not working well.
